project/Teacher.cs

```diff
@@ -69,5 +75,61 @@ public void ViewSchedule()
                 Console.WriteLine($"- {subject.SubjectName} (ID: {subject.SubjectId})");
             }
         }
+
+        private readonly List<Timeslot> _assignedTimeslots = new();
+
+        public List<Timeslot> AssignedTimeslots => new List<Timeslot>(_assignedTimeslots);
+
+        public void AssignTimeslot(Timeslot timeslot)
+        {
+            if (timeslot == null)
+                throw new ArgumentNullException(nameof(timeslot), "Timeslot cannot be null.");
+
+            if (_assignedTimeslots.Contains(timeslot))
+                throw new InvalidOperationException("This timeslot is already assigned to the teacher.");
+
+            _assignedTimeslots.Add(timeslot);
+        }
+
+        public void RemoveTimeslot(Timeslot timeslot)
+        {
+            if (timeslot == null)
+                throw new ArgumentNullException(nameof(timeslot), "Timeslot cannot be null.");
+
+            if (!_assignedTimeslots.Remove(timeslot))
+                throw new InvalidOperationException("This timeslot is not assigned to the teacher.");
+        }
+        private readonly List<Grade> _grades = new();
+
+        public List<Grade> Grades => new List<Grade>(_grades);
+
+        public void AssignGrade(Grade grade)
+        {
+            if (grade == null)
+                throw new ArgumentNullException(nameof(grade), "Grade cannot be null.");
+
+            if (_grades.Contains(grade))
+                throw new InvalidOperationException("This grade is already assigned to this teacher.");
+
+            _grades.Add(grade);
+            if (grade.Teacher != this)
```

Comment:
Again, we're adding or updating on the grade teacher despite us not allowing to add an existing grade in the previous lines?